plutus-core/cost-model/create-cost-model/CostModelCreation.hs

```diff
@@ -254,10 +342,64 @@ lessThanEqualsInteger cpuModelR = do
+appendByteString cpuModelR = do
+  cpuModel <- readModelAddedSizes cpuModelR
+  -- The buffer gets reallocated
+  let memModel = ModelTwoArgumentsAddedSizes $ ModelAddedSizes 0 1
+  pure $ CostingFun (ModelTwoArgumentsAddedSizes cpuModel) memModel
+
+
+-- The things marked with ### comments below are plausible guesses at the form
+-- of the model to get it fixed for the HF.  Later we'll have to add benchmarks
+-- to get data and R code to fit models for all of them.
+
+-- ### TODO: get model from R ###
+consByteString ::  MonadR m => (SomeSEXP (Region m)) -> m (CostingFun ModelTwoArguments)
+consByteString _ = do
+      let cpuModel = ModelAddedSizes 10000 1000  -- 10000 + 1000*(x+y)
+      let memModel = ModelTwoArgumentsAddedSizes $ ModelAddedSizes 0 1
+      pure $ CostingFun (ModelTwoArgumentsAddedSizes cpuModel) memModel
+
+-- ### TODO: get model from R ###
+sliceByteString ::  MonadR m => (SomeSEXP (Region m)) -> m (CostingFun ModelThreeArguments)
+sliceByteString _ = do
+  let cpuModel = ModelThreeArgumentsLinearInZ $ ModelLinearSize 150000 5000
+  let memModel = ModelThreeArgumentsLinearInZ $ ModelLinearSize 0 1
+  pure $ CostingFun cpuModel memModel
+-- This will be a bit tricky.  We're looking at sliceByteString p l s, which
+-- gets the substring of length l starting at p.  Probably the best we can do is
+-- to make it linear in the size of s.  We'll have to run some experiments to
+-- see which p and l give the worst case and then run benchmarks with
+-- bytestrings of various sizes and the worst case p and l for each.  For the
+-- size, the real size of the result will be bounded above by l, but the costing
+-- function only knows about the _size_ of l, so let's just make it the size of
+-- s.  That could be a big overstimate, but memory doesn't seem to be too much
+-- of a problem.
+
+
+-- ### TODO: get model from R ###
+lengthOfByteString ::  MonadR m => (SomeSEXP (Region m)) -> m (CostingFun ModelOneArgument)
+lengthOfByteString _ = do
+      let cpuModel = ModelOneArgumentConstantCost 150000 -- ### Get this from R
+      let memModel = ModelOneArgumentConstantCost 4  -- One word. Probably need some heap overhead.
+      pure $ CostingFun cpuModel memModel
+
+indexByteString ::  MonadR m => (SomeSEXP (Region m)) -> m (CostingFun ModelTwoArguments)
+indexByteString _ = do
+      let cpuModel = ModelTwoArgumentsConstantCost 150000 -- ### Get this from R
+      let memModel = ModelTwoArgumentsConstantCost 1  -- One byte. Probably need some heap overhead.
+      pure $ CostingFun cpuModel memModel
+
 equalsByteString :: MonadR m => (SomeSEXP (Region m)) -> m (CostingFun ModelTwoArguments)
 equalsByteString cpuModelR = do
-  cpuModel <- readModelMinSize cpuModelR
-  pure $ CostingFun (ModelTwoArgumentsMinSize cpuModel) boolMemModel
+  cpuModelOnDiagonal <- readModelLinear cpuModelR
+  let cpuModel = ModelTwoArgumentsLinearOnDiagonal $ ModelConstantOrLinear 150000
```

Comment:
not sure what linear on the diagonal means, but okay

Reply:
If you're comparing two bytestrings for equality (say) then it returns immediately if they have different sizes.  Otherwise in the worst case (identical inputs) it's linear in the size.  Maybe I'm trying to be too clever here and we should just take worst case behaviour in all cases.